src/proto/hybrid_vec.rs

```diff
@@ -0,0 +1,1832 @@
+use alloc::heap::{EMPTY, allocate, reallocate, deallocate};
+
+// Implemented on top of `core` to demonstrate the ability to substitute HybridVec for Vec
+use core::borrow::{BorrowFrom, ToOwned};
+use core::default::Default;
+use core::fmt;
+use core::hash::{self, Hash};
+use core::intrinsics::forget;
+use core::iter;
+use core::marker::ContravariantLifetime;
+use core::nonzero::NonZero;
+use core::mem;
+use core::num::{Int, UnsignedInt};
+use core::ops;
+use core::ptr;
+use core::raw::Slice as RawSlice;
+use core::usize;
+
+// Make this configurable?
+const ARRAY_SIZE: usize = 32;
+// The number of bytes a type has to be larger than to automatically be allocated on the heap.
+// This limits the on-stack size of `HybridVec` to a maximum of 
+// `ARRAY_SIZE * HEAP_BYTES_THRESHOLD`.
+const HEAP_BYTES_THRESHOLD: usize = 64;
+
+/// An implementation of `HybridVec` that uses stack-allocated storage below a certain length.
+/// A drop-in replacement for `Vec` from the stdlib, with a couple minor changes to the API and internal
+/// behavior.
+pub enum HybridVec<T> {
+    Stack {
+        array: [T; ARRAY_SIZE], 
```

Comment:
While elegant, this approach is actually very problematic.

By using an enum like this, if I make `T` a not tiny type then the size of `HybridVec<T>` grows dramatically, even when in Heap mode. For instance, if `T` is at the maximum `HEAP_BYTES_THRESHOLD` of 64 bytes, then `HybridVec<T>` will be at least `2048 bytes`. That's *huge* for a single struct, especially something that's meant to be a heap-allocated container that is relatively cheap to move. Considering rust's current zeroing and moving behavior, a commonly-moved `2048` byte struct can wreck the performance of a program using it.

The advantage of SSO in the way that is is usually implemented is that the `T` stored in strings is usually `u8`, and so you can fit 23 characters and a flag in the same space as is normally used for the `(ptr, len, cap)` triple, so the size of the structure is the same small 24 bytes no matter what. I really question the utility of a hybrid structure like this when it makes the heap-allocated case huge as well.